SerialPrograms/Source/PokemonSV/Programs/Farming/PokemonSV_MaterialFarmer.cpp

```diff
@@ -0,0 +1,566 @@
+                env.console.log("Move-forward iteration number: " + std::to_string(i + 1) + "/" + std::to_string(total_iterations), COLOR_PURPLE);
+
+                lets_go_movement1(context);
+            });
+        }
+    };
+
+    move_forward_with_lets_go(total_iterations);
+
+}
+
+/* 
+- This function wraps around an action (e.g. go out of PokeCenter to make a sandwich) so that
+we can handle pokemon wild encounters when executing the action.
+- Whenever a battle happens, we run away.
+- After battle ends, move back to PokeCenter to start the `action` again.
+- `action` must be an action starting at the PokeCenter 
+*/
+void MaterialFarmer::handle_battles_and_back_to_pokecenter(SingleSwitchProgramEnvironment& env, BotBaseContext& context, 
+    std::function<void(SingleSwitchProgramEnvironment& env, BotBaseContext& context)>&& action)
+{
+    bool action_finished = false;
+    bool first_iteration = true;
+    // a flag for the case that the action has finished but not yet returned to pokecenter
+    bool returned_to_pokecenter = false;
+    while(action_finished == false || returned_to_pokecenter == false){
+        // env.console.overlay().add_log("Calculate what to do next");
+        NormalBattleMenuWatcher battle_menu(COLOR_RED);
+        int ret = run_until(
+            env.console, context,
+            [&](BotBaseContext& context){
+                if (action_finished){
+                    // `action` is already finished. Now we just try to get back to pokecenter:
+                    reset_to_pokecenter(env.program_info(), env.console, context);
+                    context.wait_for_all_requests();
+                    returned_to_pokecenter = true;
+                    return;
+                }
+                // We still need to carry out `action`
+                if (first_iteration){
+                    first_iteration = false;
+                }
+                else{
+                    // This is at least the second iteration in the while-loop.
+                    // So a previous round of action failed.
+                    // We need to first re-initialize our position to the PokeCenter
+                    // Use map to fly back to the pokecenter
+                    reset_to_pokecenter(env.program_info(), env.console, context);
+                }
+                context.wait_for_all_requests();
+                action(env, context);
+                context.wait_for_all_requests();
+                action_finished = true;
+            },
+            {battle_menu}
+        );
+        if (ret == 0){
+            env.console.log("Detected battle. Now running away.", COLOR_PURPLE);
+            env.console.overlay().add_log("Detected battle. Now running away.");
+            try{
+                run_from_battle(env.program_info(), env.console, context);
+            }catch (OperationFailedException& e){
+                throw FatalProgramException(std::move(e));
+            }
```

Comment:
So if I understand the logic here correctly. If you get attacked, you fly back and try again right? So that way if you encounter a shiny (which you cannot kill), flying away will force despawn it so won't get in the way.

(I'm thinking in some later version after everything goes in, we can re-add the shiny detection + action table. lol)

Reply:
Yes. That's the logic.

If you want shiny detection, ShinyHuntScatterbug::handle_battles_and_back_to_pokecenter already does this. It's basically the same function, except I just run away by default in my version, whereas the scatterbug version uses the action table.